File: Top_100_Liked_Questions/763._Partition_Labels/solution.java
class Solution {
    public List<Integer> partitionLabels(String S) {
        // put the last occuring indices of each letter in array of size 26
        int[] lastIndices = new int[26];
        List<Integer> ans = new ArrayList<>();
        int a_ascii = (int) 'a';

        for (int i=0; i<S.length(); i++)
            lastIndices[(int) S.charAt(i) - a_ascii] = i;

        for (int i=0; i<S.length();) {
            int earliest_partition = lastIndices[(int) S.charAt(i) - a_ascii];
            int j;
            for (j=i; j<earliest_partition; j++) {
                earliest_partition = Math.max(earliest_partition, lastIndices[(int) S.charAt(j) - a_ascii]);
            }
            ans.add(j - i + 1);
            i = j + 1;
        }

        return ans;
    }
}

File: Top_100_Liked_Questions/763._Partition_Labels/solution.py
class Solution:
    def partitionLabels(self, S: str) -> List[int]:
        last_idx = [0] * 26
        ans = []

        for i,ch in enumerate(S):
            last_idx[ord(ch) - ord('a')] = i

        i = 0
        while i < len(S):
            j = i
            part = last_idx[ord(S[j]) - ord('a')]
            while j < part:
                part = max(part, last_idx[ord(S[j]) - ord('a')])
                j += 1

            ans.append(j - i + 1)
            i = j + 1

        return ans
            